scripts/word_embeddings/train_fasttext.py

```diff
@@ -0,0 +1,461 @@
+                        dim=1)
+
+                loss = loss_function(pred, label)
+
+            loss.backward()
+
+            # Normalize gradients
+            if args.groupwise_clip_gradient > 0:
+                clip_embeddings_gradients(trainer._params,
+                                          args.groupwise_clip_gradient)
+
+            if args.optimizer != 'adagrad':
+                trainer.set_learning_rate(args.lr * (1 - progress))
+            trainer.step(batch_size=1)
+
+            # Logging
+            if i % args.eval_interval == 0:
+                with print_time('mx.nd.waitall()'):
+                    mx.nd.waitall()
+
+                log(args, sw, embedding, embedding_out, loss, num_update,
+                    vocab, subword_function)
+
+        # Log at the end of every epoch
+        with print_time('mx.nd.waitall()'):
+            mx.nd.waitall()
+        log(args, sw, embedding, embedding_out, loss, num_update, vocab,
+            subword_function)
+
+        # Save params at end of epoch
+        save_params(args, embedding, embedding_out)
+
+    sw.close()
+
+
+def log(args, sw, embedding, embedding_out, loss, num_update, vocab,
+        subword_function):
+    """Logging helper"""
+    context = get_context(args)
+
+    # Word embeddings
+    embedding_norm = embedding.embedding.weight.data(
+        ctx=context[0]).as_in_context(
+            mx.cpu()).tostype('default').norm(axis=1)
+    sw.add_histogram(tag='embedding_norm', values=embedding_norm,
+                     global_step=num_update, bins=200)
+    if embedding.embedding.weight.grad(ctx=context[0]).stype == 'row_sparse':
+        embedding_grad_norm = embedding.embedding.weight.grad(
+            ctx=context[0]).data.as_in_context(
+                mx.cpu()).tostype('default').norm(axis=1)
+        sw.add_histogram(tag='embedding_grad_norm', values=embedding_grad_norm,
+                         global_step=num_update, bins=200)
+
+    # Subword embeddings
+    embedding_norm = embedding.subword_embedding.embedding.weight.data(
+        ctx=context[0]).as_in_context(
+            mx.cpu()).tostype('default').norm(axis=1)
+    sw.add_histogram(tag='subword_embedding_norm', values=embedding_norm,
+                     global_step=num_update, bins=200)
+    if embedding.subword_embedding.embedding.weight.grad(
+            ctx=context[0]).stype == 'row_sparse':
+        embedding_grad_norm = \
+            embedding.subword_embedding.embedding.weight.grad(
+                ctx=context[0]).data.as_in_context(
+                    mx.cpu()).tostype('default').norm(axis=1)
+        sw.add_histogram(tag='subword_embedding_grad_norm',
+                         values=embedding_grad_norm, global_step=num_update,
+                         bins=200)
+
+    embedding_out_norm = embedding_out.embedding.weight.data(
+        ctx=context[0]).as_in_context(
+            mx.cpu()).tostype('default').norm(axis=1)
```

Comment:
Did you implement norm(row_sparse, axis=x)? It should be straight forward to implement if we just return a 1-D dense result. It should be useful for other tasks, too